cedar-policy-core/src/authorizer/partial_response.rs

```diff
@@ -0,0 +1,617 @@
+    pub satisfied_permits: HashMap<PolicyID, Arc<Annotations>>,
+    /// All of the [`Effect::Permit`] policies that were not satisfied
+    pub false_permits: HashMap<PolicyID, (ErrorState, Arc<Annotations>)>,
+    /// All of the [`Effect::Permit`] policies that evaluated to a residual
+    pub residual_permits: HashMap<PolicyID, (Arc<Expr>, Arc<Annotations>)>,
+    /// All of the [`Effect::Forbid`] policies that were satisfied
+    pub satisfied_forbids: HashMap<PolicyID, Arc<Annotations>>,
+    /// All of the [`Effect::Forbid`] policies that were not satisfied
+    pub false_forbids: HashMap<PolicyID, (ErrorState, Arc<Annotations>)>,
+    /// All of the [`Effect::Forbid`] policies that evaluated to a residual
+    pub residual_forbids: HashMap<PolicyID, (Arc<Expr>, Arc<Annotations>)>,
+    /// All of the policy errors encountered during evaluation
+    pub errors: Vec<AuthorizationError>,
+    /// The trivial `true` expression, used for materializing a residual for satisfied policies
+    true_expr: Arc<Expr>,
+    /// The trivial `false` expression, used for materializing a residual for non-satisfied policies
+    false_expr: Arc<Expr>,
+}
+
+impl PartialResponse {
+    /// Create a partial response from each of the policy result categories
+    pub fn new(
+        true_permits: impl IntoIterator<Item = (PolicyID, Arc<Annotations>)>,
+        false_permits: impl IntoIterator<Item = (PolicyID, (ErrorState, Arc<Annotations>))>,
+        residual_permits: impl IntoIterator<Item = (PolicyID, (Arc<Expr>, Arc<Annotations>))>,
+        true_forbids: impl IntoIterator<Item = (PolicyID, Arc<Annotations>)>,
+        false_forbids: impl IntoIterator<Item = (PolicyID, (ErrorState, Arc<Annotations>))>,
+        residual_forbids: impl IntoIterator<Item = (PolicyID, (Arc<Expr>, Arc<Annotations>))>,
+        errors: impl IntoIterator<Item = AuthorizationError>,
+    ) -> Self {
+        Self {
+            satisfied_permits: true_permits.into_iter().collect(),
+            false_permits: false_permits.into_iter().collect(),
+            residual_permits: residual_permits.into_iter().collect(),
+            satisfied_forbids: true_forbids.into_iter().collect(),
+            false_forbids: false_forbids.into_iter().collect(),
+            residual_forbids: residual_forbids.into_iter().collect(),
+            errors: errors.into_iter().collect(),
+            true_expr: Arc::new(Expr::val(true)),
+            false_expr: Arc::new(Expr::val(false)),
+        }
+    }
+
+    /// Convert this response into a concrete evaluation response.
+    /// All residuals are treated as errors
+    pub fn concretize(self) -> Response {
+        self.into()
+    }
+
+    /// Attempt to reach a partial decision; the presence of residuals may result in returning [`None`],
+    /// indicating that a decision could not be reached given the unknowns
+    pub fn decision(&self) -> Option<Decision> {
+        match (
```

Comment:
I would find this code easier to read if you ordered by the "strength" of knowing any of these is empty (so roughly the order you check them below):
```
satisfied_forbids,
satisfied_permits,
residual_forbids,
residual_permits,
```